metrix/src/parser.rs

```diff
@@ -50,6 +53,77 @@ pub fn parse_query_string(input: String) -> Result<ExpressionType, &'static str>
     }
 }
 
+// This uses the proper nom 5 approach, as opposed to the outdated nom 4 approach like the rest
+// of the parser. The rest will need to be refactored.
+pub fn parse_parameter_name(input: String) -> Result<FieldType, &'static str> {
+    match param_name(&input.into_bytes()) {
+        Ok((_, o)) => Ok(o),
+        Err(_) => Err("Failed to parse parameter name"),
+    }
+}
+
+fn param_name(s: &[u8]) -> IResult<&[u8], FieldType> {
+    alt((
+        map(
+            tuple((
+                take_while1(is_sql_identifier),
+                tag("."),
+                sub_param_name,
+            )),
+            |(field, _, sub_fields)| {
+                FieldType::RootField(
+                    Field {
+                        field_root: str::from_utf8(field).unwrap().to_string(),
+                        sub_fields: Box::new(sub_fields),
+                    }
+                )
+            }
+        ),
+        map(
+            take_while1(is_sql_identifier),
+            |field| {
+                FieldType::RootField(
+                    Field {
+                        field_root: str::from_utf8(field).unwrap().to_string(),
+                        sub_fields: Box::new(FieldType::TerminalField(())),
+                    }
+                )
+            }
+        ),
+    ))(s)
+}
+
+fn sub_param_name(s: &[u8]) -> IResult<&[u8], FieldType> {
+    alt((
+        map(
+            tuple((
+                take_while1(is_sql_identifier),
+                tag("."),
+                sub_param_name,
+            )),
+            |(field, _, sub_fields)| {
+                FieldType::NestedField(
+                    Field {
+                        field_root: str::from_utf8(field).unwrap().to_string(),
+                        sub_fields: Box::new(sub_fields),
+                    }
+                )
+            }
+        ),
+        map(
+            take_while1(is_sql_identifier),
+            |field| {
+                FieldType::NestedField(
+                    Field {
+                        field_root: str::from_utf8(field).unwrap().to_string(),
+                        sub_fields: Box::new(FieldType::TerminalField(())),
+                    }
+                )
+            }
+        ),
+    ))(s)
+}
```

Comment:
parsing is a little whacky- it looks like you take each of the strings split on a period, and then assign them as `Field`s with `sub_fields` which are of type `Field`, until the final one, which is `TerminalField`?

Reply:
Yeah it parses it into a linked list. The parser is structured in a recursive way, but in reality it doesn't actually do any string splitting or anything like that, it greedily consumes characters and matches them based on how the parser was constructed.